src/page/page_cache.rs

```diff
@@ -1,50 +1,65 @@
-use super::*;
+use std::io::{Read, Seek, Write};
 
+use crossbeam::sync::AtomicOption;
 use rayon::prelude::*;
+use zstd::block::{compress, decompress};
+
+use super::*;
 
 /// A lock-free pagecache.
-pub struct PageCache<L: Log, M>
-    where M: Materializer + Sized,
-          L: Log + Sized
-{
-    t: M,
-    inner: Radix<Stack<CacheEntry<M>>>,
-    max_id: AtomicUsize,
+pub struct PageCache<PM, L, P, R> {
+    t: PM,
+    inner: Radix<Stack<CacheEntry<P>>>,
+    max_pid: AtomicUsize,
     free: Stack<PageID>,
-    log: Box<L>,
+    log: L,
     lru: lru::Lru,
+    updates: AtomicUsize,
+    last_snapshot: AtomicOption<Snapshot<R>>,
+}
+
+unsafe impl<PM, L, P, R> Send for PageCache<PM, L, P, R>
+    where PM: Send,
+          L: Send,
+          R: Send
+{
 }
 
-unsafe impl<L: Log, M: Materializer> Send for PageCache<L, M> {}
-unsafe impl<L: Log, M: Materializer> Sync for PageCache<L, M> {}
+unsafe impl<PM, L, P, R> Sync for PageCache<PM, L, P, R>
+    where PM: Sync,
+          L: Sync
+{
+}
 
-impl<L: Log, M: Materializer> Debug for PageCache<L, M> {
+impl<PM, L, P, R> Debug for PageCache<PM, L, P, R> {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
         f.write_str(&*format!(
             "PageCache {{ max: {:?} free: {:?} }}\n",
-            self.max_id.load(SeqCst),
+            self.max_pid.load(SeqCst),
             self.free
         ))
     }
 }
 
-impl<M> PageCache<LockFreeLog, M>
-    where M: Materializer,
-          M::PartialPage: Clone,
-          M: DeserializeOwned,
-          M: Serialize
+impl<PM, P, R> PageCache<PM, LockFreeLog, P, R>
+    where PM: Materializer<PartialPage = P, Recovery = R>,
+          PM: Send + Sync,
+          P: PartialEq + Clone + Debug + Serialize + DeserializeOwned + Send,
+          R: PartialEq + Clone + Debug + Serialize + DeserializeOwned
```

Comment:
I split the informatino about the Materializer's associated types because I was having trouble showing that the materializer's recovery type was deserializable before this. Sometimes splitting things out more lets the compiler see that something is actually possible.